src/Geolocator.Plugin.Android.Fused/FusedGeolocatorImplementation.cs

```diff
@@ -0,0 +1,301 @@
+						{
+							Longitude = address.Longitude,
+							Latitude = address.Latitude,
+							FeatureName = address.FeatureName,
+							PostalCode = address.PostalCode,
+							SubLocality = address.SubLocality,
+							CountryCode = address.CountryCode,
+							CountryName = address.CountryName,
+							Thoroughfare = address.Thoroughfare,
+							SubThoroughfare = address.SubThoroughfare,
+							Locality = address.Locality
+						});
+            }
+        }
+
+        public async Task<bool> StartListeningAsync(TimeSpan minimumTime, double minimumDistance, bool includeHeading = false, ListenerSettings listenerSettings = null)
+        {
+			var hasPermission = await CheckPermissions();
+			if (!hasPermission)
+				throw new GeolocationException(GeolocationError.Unauthorized);
+			
+            if (_client == null)
+			{
+				Initialize();
+			}
+
+            if (!_client.IsConnected)
+                _client.Connect();
+
+            if (!_client.IsConnected)
+                return await Task.FromResult(false);
+            
+            var minTime = (long)minimumTime.TotalMilliseconds;
+            var locationRequest = new LocationRequest();
+            locationRequest.SetSmallestDisplacement(Convert.ToSingle(minimumDistance))
+                .SetFastestInterval(minTime)
+                .SetInterval(minTime * 3)
+                .SetMaxWaitTime(minTime * 6)
+                .SetPriority(GetPriority());
+
+            var result = await LocationServices.FusedLocationApi.RequestLocationUpdatesAsync(_client, locationRequest, this);
+
+            if (result.IsSuccess)
+                IsListening = true;
+
+            return result.IsSuccess;
+        }
+
+        public async Task<bool> StopListeningAsync()
+        {
+			if (_client == null)
+				return true;
+
+            if (_client.IsConnected)
+            {
+                var result = await LocationServices.FusedLocationApi.RemoveLocationUpdatesAsync(_client, this);
+                if (!result.IsSuccess)
+                    return false;
+                
+				IsListening = false;
+            }
+
+			if (_client.IsConnected || _client.IsConnecting)
+			{
+				_client.Disconnect();
+				_client = null;
+			}
+
+            return true;
+        }
+
+        private int GetPriority()
+        {
+            if (DesiredAccuracy < 50)
+                return LocationRequest.PriorityHighAccuracy;
+
+            if (DesiredAccuracy < 100)
+                return LocationRequest.PriorityBalancedPowerAccuracy;
+
+            if (DesiredAccuracy < 200)
+                return LocationRequest.PriorityLowPower;
+
+            return LocationRequest.PriorityNoPower;
+        }
+
+        private Task<Position> NextLocationAsync()
+        {
+            var locationSource = new TaskCompletionSource<Position>();
+            EventHandler<PositionEventArgs> handler = null;
+
+            handler = (sender, args) =>
+            {
+                PositionChanged -= handler;
```

Comment:
Since I already wrote all the stuff below before seeing James' post when returning, as a reference for the time being: I fixed that issue with a slightly adjusted version of your code:

		private object syncRoot = new object();
		private TaskCompletionSource<Position> locationSource = null;

		private Task<Position> NextLocationAsync()
		{
			TaskCompletionSource<Position> localSource;
			lock (syncRoot)
			{
				//get local copy of the task completion source
				localSource = locationSource;
				if (localSource == null)
				{
					//create new tcs
					this.locationSource = localSource = new TaskCompletionSource<Position>();
				}

				void Handler(object sender, PositionEventArgs args)
				{
					PositionChanged -= Handler;
					localSource.TrySetResult(args.Position);
				}

				//in case we don't get an update at all for whatever reason,
				//abort with fallback to cached position, if any
				Task.Delay(5000).ContinueWith(t =>
				{
					localSource.TrySetResult(_lastKnownPosition);
				});

				PositionChanged += Handler;
			}

			return localSource.Task;
		}


Also, I added a conditional abort in `GetPositionAsync` by looking at the returned task result in case `StartListeningAsync`

			bool status = await StartListeningAsync(TimeSpan.FromMilliseconds(500), 10, includeHeading);
			if (!status) return _lastKnownPosition;

That along with an async connection (taken from MisterJimson, but also with an abort after 5 secs) fixed the issues mentioned before.